homeassistant/components/tesla_fleet/helpers.py

```diff
@@ -0,0 +1,78 @@
+"""Tesla Fleet helper functions."""
+
+import asyncio
+from typing import Any
+
+from tesla_fleet_api.exceptions import TeslaFleetError
+
+from homeassistant.exceptions import HomeAssistantError
+
+from .const import DOMAIN, LOGGER, TeslaFleetState
+
+
+async def wake_up_vehicle(vehicle) -> None:
+    """Wake up a vehicle."""
+    async with vehicle.wakelock:
+        times = 0
+        while vehicle.coordinator.data["state"] != TeslaFleetState.ONLINE:
+            try:
+                if times == 0:
+                    cmd = await vehicle.api.wake_up()
+                else:
+                    cmd = await vehicle.api.vehicle()
+                state = cmd["response"]["state"]
+            except TeslaFleetError as e:
+                raise HomeAssistantError(str(e)) from e
+            vehicle.coordinator.data["state"] = state
```

Comment:
>may invoke the 'wake' multiple times

That Isn't possible because I am using an asyncio lock on line 15.

Is it being outside the coordinator an issue? Its just going to be a bunch of work to refactor both this and Teslemetry to keep things consistent for maintenance..
https://github.com/home-assistant/core/blob/dev/homeassistant/components/teslemetry/helpers.py#L13C11-L13C26

Reply:
I believe that stops them from running concurrently (good), but i think it will still run twice in a row when setting the temperature and setting an hvac mode.

Reply:
Your correct that `wake_up_vehicle` will get called twice but `while vehicle.coordinator.data["state"] != TeslaFleetState.ONLINE:` will equate to false on the second run because the first was successful, did `vehicle.coordinator.data["state"] = state`, and didn't raise HomeAssistantError

Reply:
Ah of course, thanks for the correction.